preprocess.py

```diff
@@ -1,255 +1,286 @@
-            X_missing = df[(df['age_group'] == age_group) &
-                           (df['Age'].isna() == True)].copy()
-
-            X_missing = X_missing.drop('age_group', axis=1)
-            X_missing = pd.get_dummies(X_missing)
-
-            # dropping features which the imputation model wasn't trained on
-            features_to_drop = set(X_missing.columns) - set(self.trained_features[age_group])
-            features_to_drop = [item for item in features_to_drop]
-            X_missing = X_missing.drop(features_to_drop, axis=1)
-
-            # adding features which the imputation model was trained on and are missing in the test set
-            features_to_add = set(self.trained_features[age_group]) - set(X_missing.columns)
-            features_to_add = [item for item in features_to_add]
-            X_missing[features_to_add] = 0
-
-            # reordering features to match the input order into the imputation model
-            X_missing = X_missing[self.trained_features[age_group]]
-
-            y_pred = np.ceil(regressor.predict(X_missing))
-            imputed_df.loc[X_missing.index, 'Age'] = y_pred
-
-        imputed_df = self.cabin_indexes_impute(imputed_df)
-        imputed_df = self.cabin_type_impute(imputed_df)
-
-        return pd.get_dummies(imputed_df)
-
-    # averaged over the 'cabin_indexes' feature to have a single numeric value
-    # missing cells were imputed with a constant value
-    def cabin_indexes_impute(self, df: pd.DataFrame) -> pd.DataFrame:
-        def avg_cabin_index(item):
-            if type(item) == list:
-                if len(item) == 0:
-                    return self.missing_indexes_value
-            return np.mean(item)
-
-        avg_cabin_index_df = df.copy()
-        avg_cabin_index_df['cabin_indexes'] = avg_cabin_index_df['cabin_indexes'].fillna(self.missing_indexes_value)
-        avg_cabin_index_df['cabin_indexes'] = avg_cabin_index_df['cabin_indexes'].apply(avg_cabin_index)
-
-        return avg_cabin_index_df
-
-    # encoded the missingness of 'cabin_type' using an additional label - 'Missing'
-    def cabin_type_impute(self, df: pd.DataFrame) -> pd.DataFrame:
-        missingness_label = 'Missing'
-        cabin_type_missing_label_df = df.copy()
-        missing_idx = cabin_type_missing_label_df['cabin_type'][cabin_type_missing_label_df['cabin_type'].isna()].index
-
-        cabin_type_missing_label_df.loc[missing_idx, 'cabin_type'] = missingness_label
-
-        return cabin_type_missing_label_df
-
-
-# TODO: choose a method for feature selection and implement
-class SelectKBest:
-    def __init__(self):
-        super(SelectKBest, self).__init__()
-
-    def fit(self, X: pd.DataFrame, y: pd.DataFrame):
-        pass
-
-    def transform(self, X: pd.DataFrame):
-        pass
+from __future__ import annotations
+import pandas as pd
+import numpy as np
+from sklearn.ensemble import RandomForestRegressor
+
+from itertools import chain
+
+
+class Preprocessor:

```

Comment:
```suggestion
class FeatureExtractor:
```
Preprocessor is not a meaningful name of this class work... It is too general. Preprocessing also includes feature normalization, data imputation, etc.
Think of a better name..

Reply:
changed the class and file names to FeatureExtractor, sounds better than FeatureEngineering